go/vt/schemadiff/capability.go

```diff
@@ -0,0 +1,204 @@
+			// Check if only diff is change of default
+			// we temporarily remove the DEFAULT expression (if any) from both
+			// table and ALTER statement, and compare the columns: if they're otherwise equal,
+			// then the only change can be an addition/change/removal of DEFAULT, which
+			// is instant-table.
+			tableColDefinition := colStringStrippedDown(col, true, false)
+			newColDefinition := colStringStrippedDown(opt.NewColDefinition, true, false)
+			if tableColDefinition == newColDefinition {
+				return capableOf(capabilities.InstantChangeColumnDefaultFlavorCapability)
+			}
+			// Check if:
+			// 1. this an ENUM/SET
+			// 2. and the change is to append values to the end of the list
+			// 3. and the number of added values does not increase the storage size for the enum/set
+			// 4. while still not caring about a change in the default value
+			if len(col.Type.EnumValues) > 0 && len(opt.NewColDefinition.Type.EnumValues) > 0 {
+				// both are enum or set
+				if !hasPrefix(opt.NewColDefinition.Type.EnumValues, col.Type.EnumValues) {
+					return false, nil
+				}
+				// we know the new column definition is identical to, or extends, the old definition.
+				// Now validate storage:
+				if strings.EqualFold(col.Type.Type, "enum") {
+					if len(col.Type.EnumValues) <= 255 && len(opt.NewColDefinition.Type.EnumValues) > 255 {
+						// this increases the SET storage size (1 byte for up to 8 values, 2 bytes beyond)
+						return false, nil
+					}
+				}
+				if strings.EqualFold(col.Type.Type, "set") {
+					if (len(col.Type.EnumValues)+7)/8 != (len(opt.NewColDefinition.Type.EnumValues)+7)/8 {
+						// this increases the SET storage size (1 byte for up to 8 values, 2 bytes for 8-15, etc.)
+						return false, nil
+					}
+				}
+				// Now don't care about change of default:
+				tableColDefinition := colStringStrippedDown(col, true, true)
+				newColDefinition := colStringStrippedDown(opt.NewColDefinition, true, true)
+				if tableColDefinition == newColDefinition {
+					return capableOf(capabilities.InstantExpandEnumCapability)
+				}
+			}
+		}
+		return false, nil
+	default:
+		return false, nil
+	}
+}
+
+// AlterTableCapableOfInstantDDL checks if the specific ALTER TABLE is eligible to run via ALGORITHM=INSTANT, given the existing table schema and
+// the MySQL server capabilities.
+// The function is intentionally public, as it is intended to be used buy other packages, such as onlineddl.
+func AlterTableCapableOfInstantDDL(alterTable *sqlparser.AlterTable, createTable *sqlparser.CreateTable, capableOf capabilities.CapableOf) (bool, error) {
+	if capableOf == nil {
+		return false, nil
+	}
+	capable, err := capableOf(capabilities.InstantDDLFlavorCapability)
+	if err != nil {
+		return false, err
+	}
+	if !capable {
+		return false, nil
+	}
+	if alterTable.PartitionOption != nil {
+		// no INSTANT for partitions
+		return false, nil
+	}
+	if alterTable.PartitionSpec != nil {
+		// no INSTANT for partitions
+		return false, nil
+	}
```

Comment:
Any reason not to combine this into a single if with an or?

Reply:
fixed.